### PR TITLE
PIP packages

### PR DESCRIPTION
**Pull request by Yarikk26**
## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Description:

This PR adds PIP packages building functionality

### CODE REVIEW

**Comment by mojca** [comment]
```
>python -m twine upload --repository testpypi opendaq-2.1.0-cp38-cp38-linux_x86_64.whl
Uploading distributions to https://test.pypi.org/legacy/
Enter your username:
Enter your password:
Uploading opendaq-2.1.0-cp38-cp38-linux_x86_64.whl
100% ---------------------------------------- 15.7/15.7 MB • 00:03 • 5.0 MB/s
WARNING  Error during upload. Retry with the --verbose option for more details.
ERROR    HTTPError: 400 Bad Request from https://test.pypi.org/legacy/
         Binary wheel 'opendaq-2.1.0-cp38-cp38-linux_x86_64.whl' has an unsupported platform tag 'linux_x86_64'.
```

**Comment by mojca** [comment]
I'm getting
```
Build started...
1>------ Build started: Project: ZERO_CHECK, Configuration: Debug x64 ------
1>Checking File Globs
2>------ Skipped Build: Project: test_py_opendaq, Configuration: Debug x64 ------
2>Project not selected to build for this solution configuration 
3>------ Build started: Project: py_opendaq_daq, Configuration: Debug x64 ------
3>LINK : fatal error LNK1104: cannot open file 'python311.lib'
3>Done building project "py_opendaq_daq.vcxproj" -- FAILED.
```
when I try to build the package with these changes on my local machine.

**Comment by Yarikk26** [comment]
> I'm getting
> 
> ```
> Build started...
> 1>------ Build started: Project: ZERO_CHECK, Configuration: Debug x64 ------
> 1>Checking File Globs
> 2>------ Skipped Build: Project: test_py_opendaq, Configuration: Debug x64 ------
> 2>Project not selected to build for this solution configuration 
> 3>------ Build started: Project: py_opendaq_daq, Configuration: Debug x64 ------
> 3>LINK : fatal error LNK1104: cannot open file 'python311.lib'
> 3>Done building project "py_opendaq_daq.vcxproj" -- FAILED.
> ```
> 
> when I try to build the package with these changes on my local machine (VS 2022; no special configuration to explicitly declare the desired Python version).

Clean build dir. Just followed instructions to build under Windows:
- `cmake --preset "x64/msvc-22/full"`
- switch OPENDAQ_ENABLE_PYTHON_BINDINGS_TESTS to ON by GUI app without specifying exact version
- ` cmake --build . -t py_opendaq_daq` in `build\x64\msvc-22\full`
Can't reproduce:
```
  py_opendaq.vcxproj -> C:\Users\lenovo\source\repos\openDAQ\openDAQ\build\x64\msvc-22\full\bin\Debug\py_opendaq-debug.
  lib
  Building Custom Rule C:/Users/lenovo/source/repos/openDAQ/openDAQ/bindings/python/py_opendaq_daq/CMakeLists.txt
  opendaq_daq.cpp
     Creating library C:/Users/lenovo/source/repos/openDAQ/openDAQ/build/x64/msvc-22/full/bin/Debug/opendaq.lib and obj
  ect C:/Users/lenovo/source/repos/openDAQ/openDAQ/build/x64/msvc-22/full/bin/Debug/opendaq.exp
  py_opendaq_daq.vcxproj -> C:\Users\lenovo\source\repos\openDAQ\openDAQ\build\x64\msvc-22\full\bin\Debug\opendaq.cp312
  -win_amd64.pyd
```
Could you please provide more debug info? E.g. CMakeCache.txt, CMake output or at least steps to reproduce

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -39,63 +44,86 @@ jobs:
...
-          aws s3 rm s3://${{ env.s3bucket }}/${{ env.s3path }}/ --recursive --exclude '*' --include '*.exe'
-          aws s3 cp . s3://${{ env.s3bucket }}/${{ env.s3path }}/ --recursive
+          aws s3 rm "s3://${{ env.s3bucket }}/${{ env.s3path }}/" --recursive --exclude '*' --include '*.exe'
+          aws s3 cp . "s3://${{ env.s3bucket }}/${{ env.s3path }}/" --recursive
+
+      - name: Upload Wheels to S3
+        working-directory: ${{ env.wheels_path }}
+        run: |
+          aws s3 rm "s3://${{ env.s3bucket }}/${{ env.s3wheels_win_amd64_path }}" --recursive --exclude '*' --include '*.whl'
+          aws s3 cp . "s3://${{ env.s3bucket }}/${{ env.s3wheels_win_amd64_path }}" --recursive
```
Comment (JakaMohorkoDS) [inline .github/workflows/deploy.yml:126]: This should run only on push to main,

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -263,6 +293,100 @@ jobs:
...
+        with:
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: ${{ secrets.AWS_REGION }}
+
+      - name: Upload Wheels to S3
+        working-directory: ${{ env.wheels_path }}
+        run: |
+          aws s3 rm "s3://${{ env.s3bucket }}/${{ env.s3wheels_linux_x86_64_path }}" --recursive --exclude '*' --include '*.whl'
+          aws s3 cp . "s3://${{ env.s3bucket }}/${{ env.s3wheels_linux_x86_64_path }}" --recursive
```
Comment (JakaMohorkoDS) [inline .github/workflows/deploy.yml:381]: As above, should only run on main.

File: bindings/python/package/METADATA.in [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+Name: opendaq
+Version: @VERSION@
+Summary: Open Data Acquisition Platform
+Home-page: https://opendaq.com
+Author: Blueberry d.o.o.
+Author-email: opendaq@blueberry.com
+Keywords: DAQ
+Project-URL: Homepage, https://opendaq.com/
+Project-URL: Documentation, https://docs.opendaq.com/manual/
+Project-URL: Download, https://docs.opendaq.com/
```
Comment (JakaMohorkoDS) [inline bindings/python/package/METADATA.in:11]: Can we add a link to the GitHub repo here?

File: bindings/python/package/METADATA.in [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+Classifier: Development Status :: 4 - Beta
+Classifier: Intended Audience :: Developers
+Classifier: Intended Audience :: Science/Research
+Classifier: Programming Language :: Python :: 3
+Classifier: Programming Language :: Python :: 3.8
+Classifier: Programming Language :: Python :: 3.9
+Classifier: Programming Language :: Python :: 3.10
+Classifier: Programming Language :: Python :: 3.11
+Classifier: Programming Language :: Python :: 3.12
+Requires-Python: >=3.7
```
Comment (JakaMohorkoDS) [inline bindings/python/package/METADATA.in:21]: 3.8

**Review by JakaMohorkoDS — CHANGES_REQUESTED** [review]

**Review by JakaMohorkoDS — APPROVED** [review]
Approved from my side, should also wait for @mojca to review the changes.

**Comment by mojca** [comment]
Aren't these changes going to the `release/2.0` branch, so that when we release a new patch version, the wheels can easily be generated there?

I'll do some testing now.

**Review by mojca — COMMENTED** [review]
I would prefer to see the changes that modify the functionality (`CMakeLists.txt`, remove the patch for `pybind11` etc.) separate from CI changes.

It might also be nice to have `bindings/python/package` contents added as a separate commit. So I see at least 3 separate commits in place here.

I would suggest to put the changes into the release rather than the main branch. We will almost certainly want to build some bugfix releases for 2.0.x. If we can only build the wheels in the main branch, making new wheels will be more challenging. Maybe we only want to upload wheels from the main branch to AWS, but that's a different topic/question.

**Comment by Yarikk26** [comment]
> Aren't these changes going to the `release/2.0` branch, so that when we release a new patch version, the wheels can easily be generated there?
> 
> I'll do some testing now.

We decide with @JakaMohorkoDS to move to main

**Comment by Yarikk26** [comment]
> I would prefer to see the changes that modify the functionality (`CMakeLists.txt`, remove the patch for `pybind11` etc.) separate from CI changes.
> 
> It might also be nice to have `bindings/python/package` contents added as a separate commit. So I see at least 3 separate commits in place here.
> 
> I would suggest to put the changes into the release rather than the main branch. We will almost certainly want to build some bugfix releases for 2.0.x. If we can only build the wheels in the main branch, making new wheels will be more challenging. Maybe we only want to upload wheels from the main branch to AWS, but that's a different topic/question.

I thinks these changes are closely related to each other. If you separate them, you would get half working states. You can't build working Wheels for Linux without CMake changes. You can't also test without CMake changes. Yes, I could separate them but this is eventually extra work just for the same final result.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -14,6 +14,8 @@ env:
   cmake_preset: ci
   test_path: ${{ github.workspace }}/junit_reports/
   package_path: ${{ github.workspace }}/build/_packages/
+  python_version_build: 3.11
+  python_versions: 3.8 3.9 3.10 3.11 3.12
```
[the remark this answers is not in the document]
Comment (JakaMohorkoDS) [reply .github/workflows/ci.yml:18]: Manylinux is fine as it's still faster than the Windows build.

**Comment by JakaMohorkoDS** [comment]
> Aren't these changes going to the `release/2.0` branch, so that when we release a new patch version, the wheels can easily be generated there?
> 
> I'll do some testing now.

I suppose we could, but the 2.0 branch won't really be maintained/patched, so building new python packages should not be needed. It might be worth merging it to release anyhow though (if so, I apologise for having to swap targets 3 times now)

**Comment by JakaMohorkoDS** [comment]
I resolved some of the comments, but they should be considered in future PRs once the generation script and CI will be improved.

**Review by mojca — APPROVED** [review]
Apart from me not being able to build this (I have no idea how many other users might potentially be affected), it looks ok.

Some further comments:
* Yes, the split of commits is ok, thank you very much. We could slightly improve the wording to better match "When applied, this commit will <commit message>". For example something similar to "Remove no longer needed patch for pybind11" (optionally a very short explanation), "Support building Python wheel packages", "Build Python wheels as part of CI", ...
* I would still like to see the target to be `release`. If nothing else, we did in fact use these changes to build the packages that we uploaded. And even if this branch won't be maintained, if there is some ugly bug in Python that could easily be fixed, it will remain easy to rebuild the packages. (I don't really care if the CI commit ends up in release, but it would be really nice to have the other two.) If needed, I can take care of rebasing, just let me know.
* There are a few things that could be improved, but they can be improved later. Right now it's significantly better that we are able to build the packages at all, compared to having no binaries available. Even if it' still suboptimal, it can be improved in later commits.

**Pull request merged**